beacon-chain/rpc/validator/assignments.go

```diff
@@ -4,24 +4,90 @@ import (
 	"context"
 
 	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/feed"
+	statefeed "github.com/prysmaticlabs/prysm/beacon-chain/core/feed/state"
 	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
 	"github.com/prysmaticlabs/prysm/beacon-chain/core/state"
 	"github.com/prysmaticlabs/prysm/shared/bytesutil"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
-// GetDuties returns the committee assignment response from a given validator public key.
-// The committee assignment response contains the following fields for the current and previous epoch:
-//	1.) The list of validators in the committee.
-//	2.) The shard to which the committee is assigned.
-//	3.) The slot at which the committee is assigned.
-//	4.) The bool signaling if the validator is expected to propose a block at the assigned slot.
+// GetDuties returns the duties assigned to a list of validators specified
+// in the request object.
 func (vs *Server) GetDuties(ctx context.Context, req *ethpb.DutiesRequest) (*ethpb.DutiesResponse, error) {
 	if vs.SyncChecker.Syncing() {
 		return nil, status.Error(codes.Unavailable, "Syncing to latest head, not ready to respond")
 	}
+	return vs.duties(ctx, req)
+}
+
+// StreamDuties returns the duties assigned to a list of validators specified
+// in the request object via a server-side stream. The stream sends out new assignments in case
+// a chain re-org occurred.
+func (vs *Server) StreamDuties(req *ethpb.DutiesRequest, stream ethpb.BeaconNodeValidator_StreamDutiesServer) error {
+	if vs.SyncChecker.Syncing() {
+		return status.Error(codes.Unavailable, "Syncing to latest head, not ready to respond")
+	}
 
+	// We compute duties the very first time the endpoint is called.
+	res, err := vs.duties(stream.Context(), req)
+	if err != nil {
+		return status.Errorf(codes.Internal, "Could not compute validator duties: %v", err)
+	}
+	if err := stream.Send(res); err != nil {
+		return status.Errorf(codes.Internal, "Could not send response over stream: %v", err)
+	}
+
+	// We start a for loop which ticks on every epoch or a chain reorg.
+	stateChannel := make(chan *feed.Event, 1)
+	stateSub := vs.StateNotifier.StateFeed().Subscribe(stateChannel)
+	defer stateSub.Unsubscribe()
+	for {
+		select {
+		// Ticks every epoch to submit assignments to connected validator clients.
+		case <-vs.EpochTicker.C():
```

Comment:
When does this fire? If it is exactly at the epoch start time, then there will be an initial delay on receiving assignments in the first epoch slot (100ms to 500ms). Perhaps we could eagerly send assignments for the next epoch when they are guaranteed to be determined?

Reply:
That's great feedback, we might as well just listen for a new headstate update and check the current epoch number